src/shims/unix/fs.rs

```diff
@@ -1659,6 +1659,68 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             }
         }
     }
+
+    fn realpath(
+        &mut self,
+        path_op: &OpTy<'tcx, Tag>,
+        processed_path_op: &OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, Pointer<Option<Tag>>> {
+        let this = self.eval_context_mut();
+        this.assert_target_os_is_unix("realpath");
+
+        let pathname = this.read_path_from_c_str(this.read_pointer(path_op)?)?;
+
+        // Reject if isolation is enabled.
+        if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
+            this.reject_in_isolation("`realpath`", reject_with)?;
+            let eacc = this.eval_libc("EACCES")?;
+            this.set_last_error(eacc)?;
+            return Ok(Pointer::null());
+        }
+
+        let result = std::fs::canonicalize(pathname);
+        match result {
+            Ok(resolved) => {
+                let processed_ptr = this.read_pointer(processed_path_op)?;
+                let dest = if this.ptr_is_null(processed_ptr)? {
+                    // "If resolved_path is specified as NULL, then realpath() uses
+                    // malloc(3) to allocate a buffer of up to PATH_MAX bytes to hold
+                    // the resolved pathname, and returns a pointer to this buffer.  The
+                    // caller should deallocate this buffer using free(3)."
+                    // <https://man7.org/linux/man-pages/man3/realpath.3.html>
+                    this.alloc_os_str_as_c_str(resolved.as_os_str(), MiriMemoryKind::C.into())?
+                } else {
+                    processed_ptr
+                };
+
+                // Unix MAX_PATH is exposed via libc.
+                #[cfg(not(windows))]
```

Comment:
You are testing the *host* OS here, which is wrong. The relevant OS is the *target* OS. And you know that is a Unix one since you did `assert_target_os_is_unix`.

Reply:
Without this I get a compile error on the windows host in CI...it (rightly) complains about `libc` not being available.

I'm a bit confused about what code is enabled when, if you could give me a quick tl;dr that would be great! For example, I ran win and linux on my mac via `MIRI_TEST_TARGET` but it then failed in CI until I did changes like this. I don't think I have the right mental model of what is running when.

Reply:
> .it (rightly) complains about libc not being available.

Yes that is because you use the *host* libc, rather than the *target* `libc`.

Think of this as a 2-dimensional matrix: you can be using Linux/macOS/Windows as the host, and you could be using either one as target, and all 9 combinations are possible. When you are in the Miri source code and use `cfg` or `libc::foo`, you are doing host things. When you access stuff from `tcx`, like the `target.os` information or the helper function that looks up libc symbols, you are doing target things.

Reply:
Thank you! This is exactly what I needed